File: JokeConsumerKotlin/app/src/main/java/com/darryncampbell/jokeconsumerkotlin/MainActivity.kt
package com.darryncampbell.jokeconsumerkotlin

import android.app.Activity
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.content.LocalBroadcastManager
import android.support.v7.app.AppCompatActivity;
import android.view.Menu
import android.view.MenuItem
import android.widget.Button

import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*

class MainActivity : AppCompatActivity() {

    val request_id = 100;
    internal lateinit var broadcastResponseReceiver: BroadcastResponseReceiver

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        val startActivityButton = findViewById(R.id.btnViaStartActivity) as Button
        startActivityButton.setOnClickListener {
            val requestJokeIntent = Intent()
            requestJokeIntent.setClassName("com.darryncampbell.jokegenerator",
                "com.darryncampbell.jokegenerator.MainActivity")
            requestJokeIntent.putExtra("JokeRequest", "")
            startActivity(requestJokeIntent)
        }

        val startActivityForResultButton = findViewById(R.id.btnViaStartActivityForResult) as Button
        startActivityForResultButton.setOnClickListener {
            val requestJokeIntent = Intent()
            requestJokeIntent.setClassName(
                "com.darryncampbell.jokegenerator",
                "com.darryncampbell.jokegenerator.MainActivity")
            requestJokeIntent.putExtra("JokeRequest", "")
            startActivityForResult(requestJokeIntent, 0)
        }

        val sendBroadcastButton = findViewById(R.id.btnViaSendBroadcast) as Button
        sendBroadcastButton.setOnClickListener {
            val requestJokeIntent = Intent()
            requestJokeIntent.setClassName(
                "com.darryncampbell.jokegenerator",
                "com.darryncampbell.jokegenerator.JokeReceiver")
            val responseIntent = Intent(this,
                JokeConsumerBroadcastReceiver::class.java)
            val piResponse = PendingIntent.getBroadcast(
                applicationContext, 0, responseIntent, 0)
            requestJokeIntent.putExtra("callbackPI", piResponse)
            sendBroadcast(requestJokeIntent)
        }

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?)
    {
        if (requestCode == request_id)
        {
            if (resultCode == Activity.RESULT_OK)
            {
                //  Process the received Joke
                val theJoke = data?.getStringExtra("theJoke");
                output.setText("From StartActivityForResult: " + theJoke)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        //  Register the broadcast receiver used to receive UI updates from the IntentService
        val broadcastFilter = IntentFilter("LOCAL_ACTION")
        broadcastResponseReceiver = BroadcastResponseReceiver()
        val localBroadcastManager = LocalBroadcastManager.getInstance(this)
        localBroadcastManager.registerReceiver(broadcastResponseReceiver, broadcastFilter)
    }

    override fun onPause() {
        super.onPause()
        //  Unregister the broadcast receiver used to receive UI updates from the IntentService
        val localBroadcastManager = LocalBroadcastManager.getInstance(this)
        localBroadcastManager.unregisterReceiver(broadcastResponseReceiver)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        return when (item.itemId) {
            R.id.action_settings -> true
            else -> super.onOptionsItemSelected(item)
        }
    }

    inner class BroadcastResponseReceiver : BroadcastReceiver() {

        override fun onReceive(context: Context, intent: Intent) {
            var joke = ""
            if (intent.hasExtra("joke"))
                joke = intent.getStringExtra("joke")
            output.setText("From Broadcast Receiver: " + joke)
        }
    }
}

File: JokeConsumerXamarin/Joke Consumer Xamarin/MainActivity.cs
using System;
using Android.App;
using Android.Content;
using Android.OS;
using Android.Support.V4.Content;
using Android.Support.V7.App;
using Android.Views;
using Android.Widget;

namespace Joke_Consumer_Xamarin
{
    [Activity(Label = "@string/app_name", Theme = "@style/AppTheme.NoActionBar", MainLauncher = true)]
    public class MainActivity : AppCompatActivity
    {
        BroadcastResponseReceiver broadcastResponseReceiver;
        public TextView output;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            SetContentView(Resource.Layout.activity_main);
            output = FindViewById<TextView>(Resource.Id.output);

            Android.Support.V7.Widget.Toolbar toolbar = FindViewById<Android.Support.V7.Widget.Toolbar>(Resource.Id.toolbar);
            SetSupportActionBar(toolbar);

            Button btnStartActivity = FindViewById<Button>(Resource.Id.btnStartActivity);
            btnStartActivity.Click += StartActivityClicked;

            Button btnStartActivityForResult = FindViewById<Button>(Resource.Id.btnStartActivityForResult);
            btnStartActivityForResult.Click += StartActivityForResultClicked;

            Button btnSendBroadcast = FindViewById<Button>(Resource.Id.btnSendBroadcast);
            btnSendBroadcast.Click += SendBroadcastClicked;

        }

        public void StartActivityClicked(object sender, EventArgs args)
        {
            Intent requestJokeIntent = new Intent();
            requestJokeIntent.SetClassName(
                "com.darryncampbell.jokegenerator", 
                "com.darryncampbell.jokegenerator.MainActivity");
            requestJokeIntent.PutExtra("JokeRequest", "");
            StartActivity(requestJokeIntent);
        }

        public void StartActivityForResultClicked(object sender, EventArgs args)
        {
            Intent requestJokeIntent = new Intent();
            requestJokeIntent.SetClassName(
                "com.darryncampbell.jokegenerator", 
                "com.darryncampbell.jokegenerator.MainActivity");
            requestJokeIntent.PutExtra("JokeRequest", "");
            StartActivityForResult(requestJokeIntent, 100);
        }
        private void SendBroadcastClicked(object sender, EventArgs e)
        {
            Intent requestJokeIntent = new Intent();
            requestJokeIntent.SetClassName(
                "com.darryncampbell.jokegenerator", 
                "com.darryncampbell.jokegenerator.JokeReceiver");
            Intent responseIntent = new Intent(this, 
                Java.Lang.Class.FromType(typeof(JokeConsumerXamarinBroadcastReceiver)));
            PendingIntent piResponse = 
                PendingIntent.GetBroadcast(Application.Context, 0, responseIntent, 0);
            requestJokeIntent.PutExtra("callbackPI", piResponse);
            SendBroadcast(requestJokeIntent);
        }


        protected override void OnActivityResult(int requestCode, Result resultCode, Intent data)
        {
            base.OnActivityResult(requestCode, resultCode, data);

            switch (resultCode)
            {
                case Result.Ok:
                    {
                        string theJoke = data.GetStringExtra("theJoke");
                        output.Text = "From StartActivityForResult: " + theJoke;
                        break;
                    }
            }

        }

        protected override void OnResume()
        {
            base.OnResume();
            IntentFilter broadcastFilter = new IntentFilter("LOCAL_ACTION");
            broadcastResponseReceiver = new BroadcastResponseReceiver(this);
            LocalBroadcastManager localBroadcastManager = LocalBroadcastManager.GetInstance(this);
            localBroadcastManager.RegisterReceiver(broadcastResponseReceiver, broadcastFilter);
        }

        protected override void OnPause()
        {
            base.OnPause();
            LocalBroadcastManager localBroadcastManager = LocalBroadcastManager.GetInstance(this);
            localBroadcastManager.UnregisterReceiver(broadcastResponseReceiver);
        }

        public override bool OnCreateOptionsMenu(IMenu menu)
        {
            MenuInflater.Inflate(Resource.Menu.menu_main, menu);
            return true;
        }

        public override bool OnOptionsItemSelected(IMenuItem item)
        {
            int id = item.ItemId;
            if (id == Resource.Id.action_settings)
            {
                return true;
            }

            return base.OnOptionsItemSelected(item);
        }

        class BroadcastResponseReceiver : BroadcastReceiver
        {
            MainActivity _mainActivity;

            public BroadcastResponseReceiver(MainActivity activity) : base()
            {
                _mainActivity = activity;
            }

            public override void OnReceive(Context context, Intent intent)
            {
                if (intent.HasExtra("joke"))
                {
                   
                    string joke = intent.GetStringExtra("joke");
                    _mainActivity.output.Text = "From Broadcast Receiver: " + joke;
                }
            }
        }

    }
}

